### PR TITLE
[DO NOT MERGE] Compare spec and code

### PR DESCRIPTION
**Pull request by polinavino**
These are things I spotted comparing the spec with the code. Some are just questions, some might need code changes.

Questions I have about changing the formal spec : 
- should I add `outputsAttrsTooBig` check to it?
- should I replalce `checkRedeemers` with `checkScriptData`

### CODE REVIEW

File: alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs [hunk 1]
```diff
@@ -188,6 +190,7 @@ collectTwoPhaseScriptInputs pp tx utxo =
         Just script -> Right script
         Nothing -> Left (NoWitness hash)
     apply cost (sp, d, eu) script = (script, d : (valContext txinfo sp) : (getData tx utxo sp), eu, cost)
+    -- wrong order arguments, should be (context, datum, redeemer)
```
Comment (JaredCorduan) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs:193]: let's ask @michaelpj what order plutus is expecting

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs [hunk 2]
```diff
@@ -91,6 +91,7 @@ import Shelley.Spec.Ledger.UTxO
   ( UTxO (..),
     balance,
     txins,
+    -- ^^ the txins map is the one from Shelley - it should be the one implemented in Tx.hs in Alonzo (line 419)
```
Comment (JaredCorduan) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs:94]: we want `Cardano.Ledger.Alonzo.Tx (txins)`

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs [hunk 3]
```diff
@@ -384,6 +388,7 @@ utxoTransition = do
           )
           outputs
   null outputsAttrsTooBig ?! OutputBootAddrAttrsTooBig outputsAttrsTooBig
+  -- ^^ should I add this check to the spec?
```
Comment (JaredCorduan) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs:391]: that'd be great

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxow.hs [hunk 4]
```diff
@@ -282,6 +286,7 @@ alonzoStyleWitness = do
   case bodyNetID of
     SNothing -> pure ()
     SJust bid -> actualNetID == bid ?! WrongNetworkInTxBody actualNetID bid
+    -- ^^ Move network id body check to UTXO transition
```
Comment (JaredCorduan) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxow.hs:289]: I can do this one in a :microscope: PR

File: alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs [hunk 5]
```diff
@@ -277,6 +277,7 @@ transScriptPurpose (Certifying dcert) = P.Certifying (transDCert dcert)
 -- | Compute a Digest of the current transaction to pass to the script
 --   This is the major component of the valContext function.
 transTx ::
```
Comment (JaredCorduan) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs:279]: rename to `txinfo`? and maybe a reference to the spec, `Figure 6: Script Validation, cont.`

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxow.hs (hunk 4 above)
Comment (JaredCorduan) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxow.hs:289]: https://github.com/input-output-hk/cardano-ledger-specs/pull/2254

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs (hunk 2 above)
Comment (JaredCorduan) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs:94]: fixed in https://github.com/input-output-hk/cardano-ledger-specs/pull/2254

File: alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs (hunk 1 above)
Comment (michaelpj) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs:193]: order is datum, redeemer, context

Comment (JaredCorduan) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs:193]: fixed in https://github.com/input-output-hk/cardano-ledger-specs/pull/2259

File: alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs [hunk 6]
```diff
@@ -104,6 +104,7 @@ getData tx (UTxO m) sp = case sp of
 
 -- | Collect the inputs (Data, execution budget, costModel) for all twoPhase scripts.
 collectNNScriptInputs ::
+-- ^^ this is not used anywhere
```
Comment (JaredCorduan) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs:107]: removed in https://github.com/input-output-hk/cardano-ledger-specs/pull/2259

File: alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs [hunk 7]
```diff
@@ -259,14 +260,15 @@ feesOK pp tx (UTxO m) = do
   (minimumFee <= theFee) ?! FeeTooSmallUTxO minimumFee theFee
   -- Part 2
   if (getField @"totExunits" tx) == (ExUnits 0 0)
-    then pure ()
+    -- ^^ this check is wrong, it should check that txrdmrs is non-empty
```
Comment (JaredCorduan) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/Rules/Utxo.hs:263]: fixed in https://github.com/input-output-hk/cardano-ledger-specs/pull/2260

File: alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs [hunk 8]
```diff
@@ -242,6 +247,8 @@ evalScripts tx ((AlonzoScript.PlutusScript pscript, ds, units, cost) : rest) =
 --     3) The script is a twoPhase script, and the _txrdmrs Map of the TxWitness,
 --        contains Data for the script.
 checkScriptData ::
+-- ^^ This looks like it is also meant to check that the tx is carrying the right datum for a (nonnative) script
+-- this is why UTxO is needed (to get the datum hashes from the outputs)
```
Comment (JaredCorduan) [inline alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs:251]: @polinavino - I forget what we decided to do about this one. Do you remember?

Comment (polinavino) [reply alonzo/impl/src/Cardano/Ledger/Alonzo/PlutusScriptApi.hs:251]: I think it is fine (ie do nothing) - the datum check is separate

**Comment by polinavino** [comment]
I will close this, the only thing that might still need doing here is (pasted from slack) :
evaluateScripts still evaluates all scripts (though it is meant for Plutus ones only), returning True for Timelock ones, and
validateScript  instance, called in the UTXOW rule to validate only Timelock scripts, returns True for Plutus ones
are we going to leave this as is? it is not wrong but the correctness doesnt come from types...

This isn't a priority

**Comment by JaredCorduan** [comment]
This review was extremely helpful, thanks again @polinavino !

The `ValidateScript` class has not turned out to be as helpful as we had hoped. I'd love to re-think it once everything else for Alonzo is in good shape.

**Pull request closed without merging**